### ISSUE TITLE
Precision Problem when calculate longer length of tokens

### ISSUE THREAD

**Issue by rickyxie2004** [alexzhang13/flashattention2-custom-mask#18]
**Description:**
When using the Triton-based Flash Attention implementation (flash_attention_custom_mask from fa2_custom_mask) with custom masks, significant numerical discrepancies arise compared to PyTorch's native scaled_dot_product_attention when the sequence length is large (e.g., L=17776 for CogvideoX).

Inputs are in float16, which complies with Triton requirements.

Mask is a boolean tensor, properly broadcasted to shape (B, H, L, L).

The same queries, keys, values, and mask are fed into both implementations for comparison.

The discrepancy is minimal for short sequences but becomes substantial as the sequence length increases.

Minimal Reproducible Example:
```
from fa2_custom_mask import flash_attention_custom_mask
import torch
import torch.nn.functional as F

B, H, L, D = 2, 40, 17776, 64
sm_scale = 1 / (D ** 0.5)

fp16_q = torch.randn(B, H, L, D, dtype=torch.float16, device='cuda')
fp16_k = torch.randn(B, H, L, D, dtype=torch.float16, device='cuda')
fp16_v = torch.randn(B, H, L, D, dtype=torch.float16, device='cuda')

mask = torch.randint(0, 2, (B, 1, L, L), dtype=torch.bool, device='cuda')
mask = torch.broadcast_to(mask, (B, H, L, L))

out = flash_attention_custom_mask(fp16_q, fp16_k, fp16_v, mask, sm_scale)
out_real = F.scaled_dot_product_attention(
    fp16_q, fp16_k, fp16_v, attn_mask=mask, dropout_p=0.0, is_causal=False
)
```
```
ipdb> p (out-out_real).abs().max()
tensor(0.1146, device='cuda:0', dtype=torch.float16)
```

**Observed Behavior:**
out (Triton implementation) and out_real (PyTorch native) differ significantly, especially with longer sequences (In some ocassions even meet Nan). The error magnitude grows with sequence length. Suspected causes include mask handling or numerical stability issues within the Triton kernel.

Could you please help me to figure out the problem? Thanks so much.

**Comment by rickyxie2004** [comment]
Oh, I think it is my fault, I change (padding) the length of tokens to 17792 (which is multiple of 64) it works.

A ll in all, Thanks so much for your work!
